backend/src/lib.rs

```diff
@@ -1,104 +1,186 @@
-        witness: Vec<(&str, Vec<T>)>,
-        params: R,
-    ) -> Option<Proof>;
+/// Factory for backends without setup.
+struct WithoutSetupFactory<B>(PhantomData<B>);
 
-    fn generate_params<T: FieldElement>(size: usize) -> Params;
-}
+/// Factory implementation for backends without setup.
+impl<F: FieldElement, B: BackendWithoutSetup<F> + 'static> BackendFactory<F>
+    for WithoutSetupFactory<B>
+{
+    fn new_from_params(&self, degree: DegreeType) -> Box<dyn Backend<F>> {
+        Box::new(ConcreteBackendWithoutSetup(B::new(degree)))
+    }
 
-pub trait ProverWithoutParams {
-    fn prove<T: FieldElement>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-    ) -> Option<Proof>;
+    fn load_setup(&self, _input: &mut dyn io::Read) -> Result<Box<dyn Backend<F>>, Error> {
+        Err(Error::NotApplicable)
+    }
 }
 
-pub trait ProverAggregationWithParams {
-    fn prove<T: FieldElement, R1: io::Read, R2: io::Read>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-        proof: R1,
-        params: R2,
-    ) -> Proof;
-}
+/// Concrete dynamic dispatch Backend object, for backends without setup.
+struct ConcreteBackendWithoutSetup<B>(B);
 
-#[cfg(feature = "halo2")]
-pub struct Halo2Backend;
+/// Concrete implementation for backends with setup.
+impl<F: FieldElement, B: BackendWithoutSetup<F>> Backend<F> for ConcreteBackendWithoutSetup<B> {
+    fn prove(
+        &self,
+        pil: &Analyzed<F>,
+        fixed: &[(&str, Vec<F>)],
+        witness: Option<&[(&str, Vec<F>)]>,
+        prev_proof: Option<Proof>,
+        output_dir: &Path,
+    ) -> io::Result<()> {
+        self.0.prove(pil, fixed, witness, prev_proof, output_dir)
+    }
 
-#[cfg(feature = "halo2")]
-pub struct Halo2MockBackend;
+    fn write_setup(&self, _output: &mut dyn io::Write) -> Result<(), Error> {
+        Err(Error::NotApplicable)
+    }
+}
 
-#[cfg(feature = "halo2")]
-pub struct Halo2AggregationBackend;
+/// Factory for backends with setup.
+struct WithSetupFactory<B>(PhantomData<B>);
 
-#[cfg(feature = "halo2")]
-impl ProverWithParams for Halo2Backend {
-    fn prove<T: FieldElement, R: io::Read>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-        params: R,
-    ) -> Option<Proof> {
-        Some(halo2::prove_ast_read_params(pil, fixed, witness, params))
+/// Factory implementation for backends with setup.
+impl<F: FieldElement, B: BackendWithSetup<F> + 'static> BackendFactory<F> for WithSetupFactory<B> {
+    fn new_from_params(&self, degree: DegreeType) -> Box<dyn Backend<F>> {
+        Box::new(ConcreteBackendWithSetup(B::new_setup_from_params(degree)))
     }
 
-    fn generate_params<T: FieldElement>(size: usize) -> Params {
-        halo2::generate_params::<T>(size)
+    fn load_setup(&self, input: &mut dyn io::Read) -> Result<Box<dyn Backend<F>>, Error> {
```

Comment:
```suggestion
    fn new_with_setup(&self, input: &mut dyn io::Read) -> Result<Box<dyn Backend<F>>, Error> {
```

Reply:
but this isn't really loading it into a mutable object, it's creating an object from the input. So I think it should be called `new_with_something`

Reply:
renamed